### PR TITLE
Fix for `AKO VS creation fails when using HostRule with customPorts, aliases`

### PR DESCRIPTION
**Pull request by swathinsankaran**
This PR contains the following two changes.
1. Removes the hostname reference from HTTP policy
2. Modifies the datascript to remove the port and then
   route the traffic([reference](https://github.com/vmware/nsx-alb-datascript-samples-library/blob/master/availability/location_header_rewrite_with_nonstandard_port.md)).

TODO: FT fixes.

### CODE REVIEW

**Comment by swathinsankaran** [comment]
build ako

File: internal/nodes/avi_model_nodes.go [hunk 1]
```diff
@@ -1017,9 +1017,14 @@ func (v *AviHttpPolicySetNode) CalculateCheckSum() {
 	var checksum uint32
 	for _, hpp := range v.HppMap {
 		sort.Strings(hpp.Path)
-		sort.Strings(hpp.Host)
-		checksum = checksum + utils.Hash(utils.Stringify(hpp))
-
+		checksum += utils.Hash(utils.Stringify(hpp.Name)) +
```
Comment (chauhanshubham) [inline internal/nodes/avi_model_nodes.go:1020]: you can copy hpp into a new variable and remove Host from that.

Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go]: I have made changes as per the comment.

**Comment by swathinsankaran** [comment]
build ako

**Review by chauhanshubham — DISMISSED** [review]

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

File: internal/nodes/avi_model_nodes.go [hunk 2]
```diff
@@ -1077,6 +1067,38 @@ type AviHostPathPortPoolPG struct {
...
+func (v *AviHostPathPortPoolPG) GetCheckSum() uint32 {
+	// Calculate checksum and return
+	v.CalculateCheckSum()
+	return v.Checksum
+}
+
+func (v *AviHostPathPortPoolPG) CalculateCheckSum() {
+	var checksum uint32
+	sort.Strings(v.Path)
+	sort.Strings(v.Host)
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1079]: @swathinsankaran: `Host` are marked nil. I think we can skip sorting it.

File: internal/nodes/avi_model_nodes.go [hunk 3]
```diff
@@ -1077,6 +1067,38 @@ type AviHostPathPortPoolPG struct {
...
+	checksum = checksum + utils.Hash(utils.Stringify(v))
+	v.Checksum = checksum
+}
+
+func (v *AviHostPathPortPoolPG) GetNodeType() string {
+	// Calculate checksum and return
+	return "AviHostPathPortPoolPG"
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1086]: We can introduce constant here.

File: internal/nodes/avi_model_nodes.go [hunk 4]
```diff
@@ -1077,6 +1067,38 @@ type AviHostPathPortPoolPG struct {
...
+func (v *AviHostPathPortPoolPG) CalculateCheckSum() {
+	var checksum uint32
+	sort.Strings(v.Path)
+	sort.Strings(v.Host)
+	checksum = checksum + utils.Hash(utils.Stringify(v))
+	v.Checksum = checksum
+}
+
+func (v *AviHostPathPortPoolPG) GetNodeType() string {
+	// Calculate checksum and return
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1085]: Comment seems to be irrelevant to code in place.

File: internal/nodes/avi_model_nodes.go [hunk 5]
```diff
@@ -1077,6 +1067,38 @@ type AviHostPathPortPoolPG struct {
...
+	return "AviHostPathPortPoolPG"
+}
+
+func (v *AviHostPathPortPoolPG) CopyNode() AviModelNode {
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1089]: `AviHostPathPortPoolPG` is inner field of the `httpPolicyset` object.  will it be required? When will it be called?

File: internal/nodes/avi_model_nodes.go [hunk 6]
```diff
@@ -1302,7 +1324,9 @@ func (v *AviHTTPDataScriptNode) GetCheckSum() uint32 {
 func (v *AviHTTPDataScriptNode) CalculateCheckSum() {
 	// A sum of fields for this VS.
 	checksum := lib.DSChecksum(v.PoolGroupRefs, nil, false)
-	checksum = utils.Hash(fmt.Sprint(checksum) + utils.HTTP_DS_SCRIPT_MODIFIED)
+	for _, pgRefName := range v.PoolGroupRefs {
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1327]: I think DS will point to single PG. Do we need this for loop.?

File: internal/rest/avi_datascript.go [hunk 7]
```diff
@@ -159,7 +159,9 @@ func (rest *RestOperations) AviDSCacheAdd(rest_op *utils.RestOp, vsKey avicache.
 
 		// Datascript should not have a checksum
 		checksum := lib.DSChecksum(ds_cache_obj.PoolGroups, nil, false)
-		checksum = utils.Hash(fmt.Sprint(checksum) + utils.HTTP_DS_SCRIPT_MODIFIED)
+		for _, pgName := range ds_cache_obj.PoolGroups {
```
Comment (akshayhavile) [inline internal/rest/avi_datascript.go:162]: `DS` will point to single `PG`. Do we need it?

**Review by akshayhavile — COMMENTED** [review]
@swathinsankaran: Have we tested this solution on EVH? Datascripts are only get attached into insecure VS in SNI mode.

File: internal/rest/avi_datascript.go (hunk 7 above)
Comment (swathinsankaran) [reply internal/rest/avi_datascript.go:162]: Some UTs of panicked when I accessed the PoolGroup with index 0.

File: internal/nodes/avi_model_nodes.go (hunk 6 above)
Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go:1327]: yes, you are right. but some UTs panicked when I accessed it with the slice with 0th index.

File: internal/nodes/avi_model_nodes.go (hunk 2 above)
Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go:1079]: This function is getting called from other places so we cant remove the string sort of hosts.

File: internal/nodes/avi_model_nodes.go (hunk 3 above)
Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go:1086]: This function is no longer required/called hence I have removed it.

File: internal/nodes/avi_model_nodes.go (hunk 5 above)
Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go:1089]: This function is no longer required/called hence I have removed it.

**Comment by swathinsankaran** [comment]
@akshayhavile, yes I have tested the http policy changes in both sni and evh. DS changes in sni mode. Also the upgrade scenarios as suggested by Shubham.

**Comment by swathinsankaran** [comment]
build ako

File: internal/nodes/avi_model_nodes.go (hunk 2 above)
Comment (akshayhavile) [reply internal/nodes/avi_model_nodes.go:1079]: For receiver type `*AviHostPathPortPoolPG`, this will not create a problem as for each object of type `AviHostPathPortPoolPG` , AKO path should be nil as per new code changes.

File: internal/nodes/avi_model_nodes.go (hunk 6 above)
Comment (akshayhavile) [reply internal/nodes/avi_model_nodes.go:1309]: I think we should look into panic trace into depth. It could be somewhere we are missing setting up things properly.

Comment (swathinsankaran) [reply internal/nodes/avi_model_nodes.go:1309]: it is failing at the place where we are calculating the checksum of DS after dissociation of PG from DS. I will add a condition check whether the number of PGs associated is 1, if yes, include the DS.

**Comment by swathinsankaran** [comment]
build ako

**Comment by chauhanshubham** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by chauhanshubham** [comment]
build ako

**Review by chauhanshubham — APPROVED** [review]

**Review by akshayhavile — APPROVED** [review]
